validation-test/stdlib/Zip.swift

```diff
@@ -0,0 +1,196 @@
+    // conformance.
+    checkSequence(
+      test.expected.map { (OpaqueValue($0), OpaqueValue($1)) }, result,
+      stackTrace: SourceLocStack().with(test.loc), sameValue: compareElements)
+
+    // Check leftovers *after* doing checkSequence(), not before, to ensure
+    // that checkSequence() didn't force us to consume more elements than
+    // needed.
+    expectEqual(
+      test.expectedLeftoverSequence, s.map { $0.value },
+      stackTrace: SourceLocStack().with(test.loc))
+    expectEqual(
+      test.expectedLeftoverOther, other.map { $0.value },
+      stackTrace: SourceLocStack().with(test.loc))
+  }
+}
+
+ZipTests.test("Collections") {
+  typealias Element = (OpaqueValue<Int>, OpaqueValue<Int32>)
+  typealias Sequence = Zip2Sequence<
+    MinimalCollection<OpaqueValue<Int>>,MinimalCollection<OpaqueValue<Int32>>>
+  typealias Collection = Zip2Collection<
+    MinimalCollection<OpaqueValue<Int>>,MinimalCollection<OpaqueValue<Int32>>>
+
+  func compareElements(_ lhs: Element, rhs: Element) -> Bool {
+    return lhs.0.value == rhs.0.value && lhs.1.value == rhs.1.value
+  }
+
+  expectCollectionAssociatedTypes(
+    collectionType: Collection.self,
+    iteratorType: Sequence.Iterator.self,
+    subSequenceType: Slice<Collection>.self,
+    indexType: Collection.Index.self,
+    indicesType: DefaultIndices<Collection>.self)
+    
+  for test in zipTests {
+    let left = MinimalCollection<OpaqueValue<Int>>(
+      elements: test.sequence.map(OpaqueValue.init))
+    let right = MinimalCollection<OpaqueValue<Int32>>(
+      elements: test.other.map(OpaqueValue.init))
+
+    var result = zip(left, right)
+    expectType(Collection.self, &result)
+
+    checkCollection(
+      test.expected.map { (OpaqueValue($0), OpaqueValue($1)) }, 
+      result, sameValue: compareElements)      
+  }
+}
+
+ZipTests.test("RandomAccessCollections") {
+  typealias Element = (OpaqueValue<Int>, OpaqueValue<Int32>)
+  typealias Sequence = Zip2Sequence<
+    MinimalRandomAccessCollection<OpaqueValue<Int>>,
+    MinimalRandomAccessCollection<OpaqueValue<Int32>>>
+  typealias Collection = Zip2Collection<
+    MinimalRandomAccessCollection<OpaqueValue<Int>>,
+    MinimalRandomAccessCollection<OpaqueValue<Int32>>>
+
+  func compareElements(_ lhs: Element, rhs: Element) -> Bool {
+    return lhs.0.value == rhs.0.value && lhs.1.value == rhs.1.value
+  }
+      
+  expectCollectionAssociatedTypes(
+    collectionType: Collection.self,
+    iteratorType: Sequence.Iterator.self,
+    subSequenceType: Slice<Collection>.self,
+    indexType: Collection.Index.self,
+    indicesType: DefaultIndices<Collection>.self)
+
+  for test in zipTests {
+    let left = MinimalRandomAccessCollection<OpaqueValue<Int>>(
+      elements: test.sequence.map(OpaqueValue.init))
+    let right = MinimalRandomAccessCollection<OpaqueValue<Int32>>(
+      elements: test.other.map(OpaqueValue.init))
+
+    var result = zip(left, right)
+    expectType(Collection.self, &result)
```

Comment:
`expectType(RandomAccessCollection.self, &result)`

Reply:
Don't think so. That's `Collection`, the type alias a few lines above, rather than `Collection` the protocol. Probably best to rename that alias...